adapters/repos/db/migrator.go

```diff
@@ -113,6 +116,139 @@ func (m *Migrator) UpdateClass(ctx context.Context, className string, newClassNa
 	return nil
 }
 
+// UpdateIndex ensures that the local index is up2date with the latest sharding
+// state (shards/tenants) and index properties that may have been added in the
+// case that the local node was down during a class update operation.
+//
+// This method is relevant when the local node is a part of a cluster,
+// particularly with the introduction of the v2 RAFT-based schema
+func (m *Migrator) UpdateIndex(ctx context.Context, incomingClass *models.Class,
+	incomingSS *sharding.State,
+) error {
+	idx := m.db.GetIndex(schema.ClassName(incomingClass.Class))
+
+	{ // add index if missing
+		if idx == nil {
+			if err := m.AddClass(ctx, incomingClass, incomingSS); err != nil {
+				return fmt.Errorf(
+					"add missing class %s during update index: %w",
+					incomingClass.Class, err)
+			}
+			return nil
+		}
+	}
+
+	{ // add/remove missing shards
+		if incomingSS.PartitioningEnabled {
+			if err := m.updateIndexTenants(ctx, idx, incomingClass, incomingSS); err != nil {
+				return err
+			}
+		} else {
+			if err := m.updateIndexAddMissingShards(ctx, idx, incomingClass, incomingSS); err != nil {
+				return err
+			}
+		}
+	}
+
+	{ // add missing properties
+		if err := m.updateIndexAddMissingProperties(ctx, idx, incomingClass); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (m *Migrator) updateIndexTenants(ctx context.Context, idx *Index,
+	incomingClass *models.Class, incomingSS *sharding.State,
+) error {
+	if err := m.updateIndexAddTenants(ctx, idx, incomingClass, incomingSS); err != nil {
+		return err
+	}
+	return m.updateIndexDeleteTenants(ctx, idx, incomingClass, incomingSS)
+}
+
+func (m *Migrator) updateIndexAddTenants(ctx context.Context, idx *Index,
+	incomingClass *models.Class, incomingSS *sharding.State,
+) error {
+	for name, phys := range incomingSS.Physical {
+		// Only load the tenant if activity status == HOT
+		if schemaUC.IsLocalActiveTenant(&phys, m.db.schemaGetter.NodeName()) {
+			loaded := idx.shards.Load(name)
+			if loaded == nil {
+				if err := idx.initAndStoreShard(ctx, name, incomingClass, m.db.promMetrics); err != nil {
+					return fmt.Errorf("add missing tenant shard %s during update index: %w", name, err)
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (m *Migrator) updateIndexDeleteTenants(ctx context.Context, idx *Index,
+	incomingClass *models.Class, incomingSS *sharding.State,
+) error {
+	var (
+		localSS  = idx.getSchema.CopyShardingState(incomingClass.Class)
```

Comment:
I believe we should avoid `dx.getSchema`  because it will return the same sharding state `incomingSS`.
Is it possible to achieve the same goal without querying the schema?
I mean similar to how this has been resolved previously in `updateIndexAddTenants`

Reply:
Ah, of course. Fixed